### PR TITLE
Bump go version and all direct dependencies to newest for k8s compatibility

### LINKED ISSUE #2829: actions-runner-controller and gha-runner-scale-set-controller Go library incompatible with Kubernetes 1.27 resulting in failure to start
### Checks

- [X] I've already read https://github.com/actions/actions-runner-controller/blob/master/TROUBLESHOOTING.md and I'm sure my issue is not covered in the troubleshooting guide.
- [X] I'm not using a custom entrypoint in my runner image

### Controller Version

0.27.4 for actions-runner-controller and 0.5.0 for gha-runner-scale-set-controller

### Helm Chart Version

0.23.4 for actions-runner-controller and 0.5.0 for gha-runner-scale-set-controller

### CertManager Version

1.12.3

### Deployment Method

Helm

### cert-manager installation

- Yes
- Yes

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue
- [X] I've migrated to the workflow job webhook event (if you using webhook driven scaling)

### Resource Definitions

```yaml
Not relevant because the actions-runner-controller does not start
```


### To Reproduce

```markdown
1) Have any Kubernetes 1.27 cluster.
2) Install the actions-runner-controller in any way you see fit.
3) Watch the following errors in the describe section roll on in and the pod fail to start.
```


### Describe the bug

```
2023-08-22T19:14:29Z    INFO    Using default AutoscalingListener image pull policy    {"ImagePullPolicy": "IfNotPresent"}                                                                                                                                                           │
│ panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                              │
│ [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12798d0]                                                                                                                                                                                                              │
│                                                                                                                                                                                                                                                                                      │
│ goroutine 1 [running]:                                                                                                                                                                                                                                                               │
│ k8s.io/client-go/discovery.convertAPIResource(...)                                                                                                                                                                                                                                   │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:88                                                                                                                                                                                                                    │
│ k8s.io/client-go/discovery.convertAPIGroup({{{0x0, 0x0}, {0x0, 0x0}}, {{0xc0001cd050, 0x15}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})                                                                                                                                                     │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:69 +0x5f0                                                                                                                                                                                                             │
│ k8s.io/client-go/discovery.SplitGroupsAndResources({{{0xc0000485e8, 0x15}, {0xc000050f80, 0x1b}}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})                                                                                                                                   │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:35 +0x2f8                                                                                                                                                                                                             │
│ k8s.io/client-go/discovery.(*DiscoveryClient).downloadAPIs(0x1976251?)                                                                                                                                                                                                               │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:310 +0x47c                                                                                                                                                                                                                │
│ k8s.io/client-go/discovery.(*DiscoveryClient).GroupsAndMaybeResources(0xc00021b360?)                                                                                                                                                                                                 │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:198 +0x5c                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.ServerGroupsAndResources({0x1c0ac30, 0xc0003d3b00})                                                                                                                                                                                                       │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:392 +0x59                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.(*DiscoveryClient).ServerGroupsAndResources.func1()                                                                                                                                                                                                       │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:356 +0x25                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.withRetries(0x2, 0xc000174e10)                                                                                                                                                                                                                            │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:621 +0x72                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.(*DiscoveryClient).ServerGroupsAndResources(0x0?)                                                                                                                                                                                                         │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:355 +0x3a                                                                                                                                                                                                                 │
│ k8s.io/client-go/restmapper.GetAPIGroupResources({0x1c0ac30?, 0xc0003d3b00?})                                                                                                                                                                                                        │
│     k8s.io/client-go@v0.26.2/restmapper/discovery.go:148 +0x42                                                                                                                                                                                                                       │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.NewDynamicRESTMapper.func1()                                                                                                                                                                                                       │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:94 +0x25                                                                                                                                                                                          │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.(*dynamicRESTMapper).setStaticMapper(...)                                                                                                                                                                                          │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:130                                                                                                                                                                                               │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.NewDynamicRESTMapper(0xc00003e360?, {0x0, 0x0, 0xf689557bcccc2f01?})                                                                                                                                                               │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:110 +0x182                                                                                                                                                                                        │
│ sigs.k8s.io/controller-runtime/pkg/cluster.setOptionsDefaults.func1(0xc000190700?)                                                                                                                                                                                                   │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/cluster/cluster.go:217 +0x25                                                                                                                                                                                                          │
│ sigs.k8s.io/controller-runtime/pkg/cluster.New(0xc0001b6240, {0xc0005517b8, 0x1, 0x0?})                                                                                                                                                                                              │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/cluster/cluster.go:159 +0x18d                                                                                                                                                                                                         │
│ sigs.k8s.io/controller-runtime/pkg/manager.New(_, {0xc000190700, 0x0, 0xc000540078, {{0x1c07860, 0xc00052a080}, 0x0}, 0x1, {0x0, 0x0}, ...})                                                                                                                                         │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/manager/manager.go:359 +0xf9                                                                                                                                                                                                          │
│ main.main()                                                                                                                                                                                                                                                                          │
│     ./main.go:181 +0x123d                                                                                                                                                                                                                                                            │
│ Stream closed EOF for actions-runner-system/actions-runner-system-actions-runner-controller-7bfcb678454rxnh (manager)
```

### Describe the expected behavior

The pod starts successfully without throwing go-library related segmentation fault errors.

### Whole Controller Logs

```shell
2023-08-22T19:14:29Z    INFO    Using default AutoscalingListener image pull policy    {"ImagePullPolicy": "IfNotPresent"}                                                                                                                                                           │
│ panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                              │
│ [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12798d0]                                                                                                                                                                                                              │
│                                                                                                                                                                                                                                                                                      │
│ goroutine 1 [running]:                                                                                                                                                                                                                                                               │
│ k8s.io/client-go/discovery.convertAPIResource(...)                                                                                                                                                                                                                                   │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:88                                                                                                                                                                                                                    │
│ k8s.io/client-go/discovery.convertAPIGroup({{{0x0, 0x0}, {0x0, 0x0}}, {{0xc0001cd050, 0x15}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})                                                                                                                                                     │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:69 +0x5f0                                                                                                                                                                                                             │
│ k8s.io/client-go/discovery.SplitGroupsAndResources({{{0xc0000485e8, 0x15}, {0xc000050f80, 0x1b}}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...})                                                                                                                                   │
│     k8s.io/client-go@v0.26.2/discovery/aggregated_discovery.go:35 +0x2f8                                                                                                                                                                                                             │
│ k8s.io/client-go/discovery.(*DiscoveryClient).downloadAPIs(0x1976251?)                                                                                                                                                                                                               │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:310 +0x47c                                                                                                                                                                                                                │
│ k8s.io/client-go/discovery.(*DiscoveryClient).GroupsAndMaybeResources(0xc00021b360?)                                                                                                                                                                                                 │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:198 +0x5c                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.ServerGroupsAndResources({0x1c0ac30, 0xc0003d3b00})                                                                                                                                                                                                       │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:392 +0x59                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.(*DiscoveryClient).ServerGroupsAndResources.func1()                                                                                                                                                                                                       │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:356 +0x25                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.withRetries(0x2, 0xc000174e10)                                                                                                                                                                                                                            │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:621 +0x72                                                                                                                                                                                                                 │
│ k8s.io/client-go/discovery.(*DiscoveryClient).ServerGroupsAndResources(0x0?)                                                                                                                                                                                                         │
│     k8s.io/client-go@v0.26.2/discovery/discovery_client.go:355 +0x3a                                                                                                                                                                                                                 │
│ k8s.io/client-go/restmapper.GetAPIGroupResources({0x1c0ac30?, 0xc0003d3b00?})                                                                                                                                                                                                        │
│     k8s.io/client-go@v0.26.2/restmapper/discovery.go:148 +0x42                                                                                                                                                                                                                       │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.NewDynamicRESTMapper.func1()                                                                                                                                                                                                       │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:94 +0x25                                                                                                                                                                                          │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.(*dynamicRESTMapper).setStaticMapper(...)                                                                                                                                                                                          │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:130                                                                                                                                                                                               │
│ sigs.k8s.io/controller-runtime/pkg/client/apiutil.NewDynamicRESTMapper(0xc00003e360?, {0x0, 0x0, 0xf689557bcccc2f01?})                                                                                                                                                               │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/client/apiutil/dynamicrestmapper.go:110 +0x182                                                                                                                                                                                        │
│ sigs.k8s.io/controller-runtime/pkg/cluster.setOptionsDefaults.func1(0xc000190700?)                                                                                                                                                                                                   │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/cluster/cluster.go:217 +0x25                                                                                                                                                                                                          │
│ sigs.k8s.io/controller-runtime/pkg/cluster.New(0xc0001b6240, {0xc0005517b8, 0x1, 0x0?})                                                                                                                                                                                              │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/cluster/cluster.go:159 +0x18d                                                                                                                                                                                                         │
│ sigs.k8s.io/controller-runtime/pkg/manager.New(_, {0xc000190700, 0x0, 0xc000540078, {{0x1c07860, 0xc00052a080}, 0x0}, 0x1, {0x0, 0x0}, ...})                                                                                                                                         │
│     sigs.k8s.io/controller-runtime@v0.14.4/pkg/manager/manager.go:359 +0xf9                                                                                                                                                                                                          │
│ main.main()                                                                                                                                                                                                                                                                          │
│     ./main.go:181 +0x123d                                                                                                                                                                                                                                                            │
│ Stream closed EOF for actions-runner-system/actions-runner-system-actions-runner-controller-7bfcb678454rxnh (manager)
```
```


### Whole Runner Pod Logs

```shell
None available because the actions-runner-controller will not start!
```


### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by nikola-jokic**
Fixes #2829

### CODE REVIEW

File: .github/workflows/global-run-codeql.yaml [hunk 1]
```diff
@@ -27,6 +27,11 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v3
 
+      - name: Install Go
+        uses: actions/setup-go@v4
+        with:
+          go-version-file: go.mod
```
Comment (fhammerl) [inline .github/workflows/global-run-codeql.yaml:33]: default go version on ubuntu-latest is still 1.20 as of 7 Nov 2023

File: Dockerfile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Build the manager binary
-FROM --platform=$BUILDPLATFORM golang:1.20.7 as builder
+FROM --platform=$BUILDPLATFORM golang:1.21 as builder
```
Comment (fhammerl) [inline Dockerfile]: NIT: why not pin it like in go.mod?

File: cmd/githubwebhookserver/main.go [hunk 3]
```diff
@@ -149,11 +152,19 @@ func main() {
...
+		Cache: cache.Options{
+			SyncPeriod: &syncPeriod,
+			DefaultNamespaces: map[string]cache.Config{
+				watchNamespace: {},
+			},
+		},
+		Metrics: metricsserver.Options{
+			BindAddress: metricsAddr,
+		},
+		WebhookServer: webhook.NewServer(webhook.Options{
```
Comment (fhammerl) [inline cmd/githubwebhookserver/main.go:165]: From https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.0/pkg/manager#Options :

```
// Port is the port that the webhook server serves at.
	// It is used to set webhook.Server.Port if WebhookServer is not set.
	//
	// Deprecated: Use WebhookServer instead. A WebhookServer can be created via webhook.NewServer.
	Port [int](https://pkg.go.dev/builtin#int)
```

**Comment by fhammerl** [comment]
All but one changes to direct dependencies in `go.mod` are minor or patch bumps
Exception: `github.com/evanphx/json-patch v4.12.0+incompatible --> v5.7.0+incompatible`

Some packages are still in `v0` so a minor upgrade brought api changes, meaning refactoring (see `*.go` files for details)
Most notable upgrade bringing API changes leading to the refactoring:
`sigs.k8s.io/controller-runtime v0.14.4` -> `sigs.k8s.io/controller-runtime v0.16.3`

Should we consider breaking this up into smaller PRs for subsets of these packages?

Could you describe the `*.yaml` changes?

**Comment by nikola-jokic** [comment]
Breaking it into multiple PRs would still cause this PR to be big as we are intending to update the framework and the surrounding tooling :relaxed: 

`*yaml` changes are due to underlying resources change (`pod.containers.resizePolicy` is added)

**Review by Link- — APPROVED** [review]

**Pull request merged**